generate subscriber worker classes with more deterministic names

heythisisnate opened the pull request:
_instead of_ `Reactor::StaticSubscribers::EventNameHandler` now the dynamically generated class would be namespaced with the containing class name: `Reactor::StaticSubscribers::MySubscriberClass::EventNameHandler`

In the case of more than one event handler named exactly the same, an exception will be raised at load time. Developers should pass a `:handler_name` option on the event handler declaration to disambiguate.

This is a breaking change ... so Reactor 1.0?

bot review me @wnadeau 
cc @dougbarth 

[fixes #115182193]

wnadeau commented on lib/reactor/version.rb:
```diff
@@ -1,3 +1,3 @@
 module Reactor
-  VERSION = "0.10.1"
+  VERSION = "1.0.0"
```
i've been holding out on a `1.0` for a release that has either a compelling grand vision or a "discrete component" simplification of this thing. I think picking either direction and making it a reality is a necessity for calling this "1.0".

Would you mind making it `0.11.0`? I'm happy to consider otherwise, and I do anticipate being able to release a 1.0 sometime in Q2 (hopefully the next hackathon), as I am now leaning toward a "discrete component" simplification of Reactor (whereby several vestigial features get cut to reduce compleity)

wnadeau commented on lib/reactor/models/concerns/subscribable.rb:
```diff
@@ -43,7 +47,21 @@ def self.perform_where_needed(data)
 
-      Reactor::StaticSubscribers.const_set(new_class, klass)
+      # dynamically define a module namespace based on the subscriber's class name
+      unless Reactor::StaticSubscribers.const_defined?(source_class, false)
```
mind extracting these into descriptive private functions and nixing the comment?

wnadeau commented on lib/reactor/models/concerns/subscribable.rb:
```diff
@@ -43,7 +47,21 @@ def self.perform_where_needed(data)
+      namespace = Reactor::StaticSubscribers.const_get(source_class, false)
+
+      if namespace.const_defined?(handler_name)
+        raise EventHandlerAlreadyDefined.new %{ A Reactor event named #{handler_name} already has been defined on #{namespace}.
```
:sparkles: 
might be worth extracting a descriptive private method here too

wnadeau commented on lib/reactor/models/concerns/subscribable.rb:
```diff
@@ -12,12 +14,14 @@ def on_event(*args, &block)
-        new_class = "#{handler_class_prefix}Handler#{i}"
-        i+= 1
-      end while Reactor::StaticSubscribers.const_defined?(new_class)
+      source_class          = options[:source] ? options[:source].name : 'Anonymous'
```
source.... it looks like this would create an `Anonymous` namespace in anyone's app.

Hmmm, when is `source` nil? Searching for `StaticSubscriberFactory.create` yields only the `on_event` usage and then a couple of usages for test setup.

I'll go ahead and bet that you threw this in there to get a test to pass because it was blowing up on `source` being nil.
If so, what do you think about updating ~line 64 of `event_spec.rb` to pass a source through (to get the tests to pass) rather than falling back to creating `Anonymous` module in consuming apps?

wnadeau commented:
this is awesome!! really great idea to let people pass their own handler name if they need to

heythisisnate commented:
@wnadeau care to take another look? I addressed your feedback and also cleaned up the `Reactor::Subscribable` concern quite a bit. I removed the `StaticSubscriberFactory` class in favor of just  some private methods mixed in with `Reactor::Subscribable`. This makes some things simpler (i.e. eliminated the need to pass `source` around cause now it's just `self`). Also changed the version to 0.11.0 as requested.

dougbarth commented:
What do you plan to do about the inflight jobs that use the old naming scheme when this change goes out? After this change is deployed, Sidekiq will be unable to find those generated worker classes and throw an error, no?

heythisisnate commented:
@dougbarth That's correct. I thought about this, and I think the simplest thing to do is a short downtime deploy:

1. put up a maintenance page so users can't use the app
2. wait for sidekiq queue to go down to zero (this should only take seconds)
3. deploy new code

Anything in the retry queue referencing the old worker names would still cause issues, but I'm going to try to clear those out beforehand.

wnadeau commented:
:sparkles: :sparkles:
